### PR TITLE
TEP-0118: Add validation for matrix include pipeline parameter variables 

### PR DESCRIPTION
**Pull request by EmmaMunley**
[[TEP-0090: Matrix](https://github.com/tektoncd/community/blob/main/teps/0090-matrix.md)] introduced `Matrix` to the `PipelineTask` specification such that the `PipelineTask` executes a list of `TaskRuns` or `Runs` in parallel with the specified list of inputs for a `Parameter` or with different combinations of the inputs for a set of `Parameters`.

To build on this [TEP-0018 ](https://github.com/tektoncd/community/blob/main/teps/0118-matrix-with-explicit-combinations-of-parameters.md )introduced Matrix.Include, which allows passing in a specific combinations of `Parameters` into the `Matrix`.

**This PR validates that matrix pipeline parameter variables  including include params that may contain the reference(s) to other params are used appropriately.**

Note: This is still in preview mode. Other forms of validation and implementation logic will be added in subsequent commits.

/kind feature

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [x] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) included if any changes are user facing
- [x] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including
  functionality, content, code)
- [x] Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [x] Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings)
- [x] Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

```release-note
NONE
```

### CODE REVIEW

File: pkg/apis/pipeline/v1/param_types.go
Comment (jerop) [inline pkg/apis/pipeline/v1/param_types.go:328]: please add tests for these changes

**Review by jerop — CHANGES_REQUESTED** [review]
please add tests for validating pipeline parameter variables in matrix.include

File: pkg/apis/pipeline/v1/param_types.go
Comment (EmmaMunley) [reply pkg/apis/pipeline/v1/param_types.go:328]: I believe the tests aded in this PR: [TEP-0118: Add validation for matrix pipeline context parameter variables](https://github.com/tektoncd/pipeline/pull/6238/files#top) should cover this. Just waiting for a /lgtm and then can merge the changes.

Comment (jerop) [reply pkg/apis/pipeline/v1/param_types.go:328]: ok, will put a hold on this until the other PR is merged

/hold

Comment (EmmaMunley) [reply pkg/apis/pipeline/v1/param_types.go:328]: The tests have been added. 

/hold cancel

Comment (jerop) [reply pkg/apis/pipeline/v1/param_types.go:328]: /unhold

File: pkg/apis/pipeline/v1/param_types.go
Comment (jerop) [inline pkg/apis/pipeline/v1/param_types.go:319]: you can also make this a member function of matrix as @pritidesai suggested in other pull requests

```suggestion
func (matrix *Matrix) validatePipelineParametersVariablesInMatrixParameters(prefix string, paramNames sets.String, arrayParamNames sets.String, objectParamNameKeys map[string][]string) (errs *apis.FieldError) {
```

File: pkg/apis/pipeline/v1/pipeline_validation_test.go
Comment (jerop) [inline pkg/apis/pipeline/v1/pipeline_validation_test.go:2555]: this test is for invalid contexts everywhere, please don't remove the existing tests for context variables in non-matrix fields

File: pkg/apis/pipeline/v1/pipeline_validation_test.go
Comment (jerop) [inline pkg/apis/pipeline/v1/pipeline_validation_test.go:2560]: don't remove this test

same applies to the other test cases

File: pkg/apis/pipeline/v1/param_types.go
Comment (EmmaMunley) [reply pkg/apis/pipeline/v1/param_types.go]: This has been refactored.

File: pkg/apis/pipeline/v1/pipeline_validation_test.go
Comment (EmmaMunley) [reply pkg/apis/pipeline/v1/pipeline_validation_test.go:1636]: That was accidental.

File: pkg/apis/pipeline/v1/pipeline_validation_test.go
Comment (EmmaMunley) [reply pkg/apis/pipeline/v1/pipeline_validation_test.go]: Sorry, this was an outstanding comment from a previous PR to isolate the functionality of context variables in the matrix since it was testing both valid context variables in matrix and valid context variables in params.

 I have added an additional set of tests TestContextValidMatrix() and TestContextInvalidMatrix() to isolate the context variables in the matrix by removing Task.Params and kept the existing TestContextValid() and TestContextInalid()

File: pkg/apis/pipeline/v1/param_types.go
Comment (lbernick) [inline pkg/apis/pipeline/v1/param_types.go:320]: nit: I suggest

```
if matrix == nil {
return nil
}
```

this reduces the level of nested conditionals

File: pkg/apis/pipeline/v1/pipeline_validation_test.go
Comment (lbernick) [inline pkg/apis/pipeline/v1/pipeline_validation_test.go:1143]: can this test case be simplified? (e.g. fewer combinations)

File: pkg/apis/pipeline/v1/pipeline_validation_test.go
Comment (lbernick) [inline pkg/apis/pipeline/v1/pipeline_validation_test.go:1131]: I'm confused what this is supposed to test-- it looks like it's testing valid string parameters in matrix.include (and also has some unrelated array params), but the test name sounds like it's testing array params in matrix.params. Can you update the test name and remove the unrelated array params?

File: pkg/apis/pipeline/v1/pipeline_validation_test.go
Comment (lbernick) [inline pkg/apis/pipeline/v1/pipeline_validation_test.go:2684]: It looks like this might contain unrelated changes from #6238, can you please remove them?

File: pkg/apis/pipeline/v1/pipeline_validation_test.go
Comment (lbernick) [inline pkg/apis/pipeline/v1/pipeline_validation_test.go:1641]: What does the index here refer to? In this example, include.params[1] is named "b-param", and refers to a value that doesn't exist. I think this error message might be hard for users to understand and fix; it might make more sense to return `matrix.include.params[1]` or `matrix.include.params[1].value` instead.

(I think `params[b-param]` is a bit confusing, since params isn't a dict, but probably most people could figure this out, but `value[1]` is very confusing because value is not an array.)

File: pkg/apis/pipeline/v1/pipeline_validation_test.go
Comment (EmmaMunley) [reply pkg/apis/pipeline/v1/pipeline_validation_test.go:2684]: Oh I thought you wanted me to include these changes here since the #6238 PR merged already. And @jerop agreed we could add the TestContextValidMatrix() and TestContextValidMatrix() with the removed Task Params.

Comment (lbernick) [reply pkg/apis/pipeline/v1/pipeline_validation_test.go:2684]: Maybe the rebase didn't work quite correctly? It looks like you're adding new test classes which probably isn't what you want

Comment (EmmaMunley) [reply pkg/apis/pipeline/v1/pipeline_validation_test.go:2684]: @lbernick These test cases were never added to the PR before it was merged so I thought I would add them here?! If not I can open a separate PR for this.

From chat:
would you mind addressing my review on https://github.com/tektoncd/pipeline/pull/6238 even though it got merged? the suggestion was to remove pipelinetask.params from the test cases, since the test cases are designed to just test validation of matrix.include

Comment (lbernick) [reply pkg/apis/pipeline/v1/pipeline_validation_test.go:2684]: Thanks, sorry, I appreciate you addressing these comments! Yes, I think a separate PR would be better, where you can use the commit message and PR title to explain the reasoning for the changes. Otherwise, it's a bit confusing to understand why these changes are bundled in. (As you can see by me confusing myself :) )

File: pkg/apis/pipeline/v1/pipeline_validation_test.go
Comment (EmmaMunley) [reply pkg/apis/pipeline/v1/pipeline_validation_test.go:1141]: Simplified

File: pkg/apis/pipeline/v1/pipeline_validation_test.go
Comment (EmmaMunley) [reply pkg/apis/pipeline/v1/pipeline_validation_test.go:2667]: Okay no problem. I've created a separate PR for this here: #6279

File: pkg/apis/pipeline/v1/pipeline_validation_test.go
Comment (EmmaMunley) [reply pkg/apis/pipeline/v1/pipeline_validation_test.go]: Updated error messaging

**Comment by EmmaMunley** [comment]
Not sure why code coverage went down for an unchanged function:

```// SetDefaults set the default type
func (pp *ParamSpec) SetDefaults(context.Context) {
        if pp == nil {
                return
        }```

File: pkg/apis/pipeline/v1beta1/pipeline_validation_test.go
Comment (pritidesai) [inline pkg/apis/pipeline/v1beta1/pipeline_validation_test.go:3561]: this test is missing from `v1/pipeline_validation_test.go`, should we include it in there too?

File: pkg/apis/pipeline/v1/pipeline_validation_test.go
Comment (pritidesai) [inline pkg/apis/pipeline/v1/pipeline_validation_test.go:1131]: we might want to add a similar test for matrix.params if it does not already exist:

```
		name: "valid string parameter variables in matrix params",
		params: []ParamSpec{{
			Name: "baz", Type: ParamTypeString,
		}},
		tasks: []PipelineTask{{
			Name:    "bar",
			TaskRef: &TaskRef{Name: "bar-task"},
			Matrix: &Matrix{
				Params: []Param{{
					Name: "a-param", Value: ParamValue{Type: ParamTypeString, StringVal: "$(params.baz)"}},
				}},
		}},
```

File: pkg/apis/pipeline/v1beta1/pipeline_validation_test.go
Comment (EmmaMunley) [reply pkg/apis/pipeline/v1beta1/pipeline_validation_test.go:3561]: Thanks for the catch! This has been added.

File: pkg/apis/pipeline/v1/pipeline_validation_test.go
Comment (EmmaMunley) [reply pkg/apis/pipeline/v1/pipeline_validation_test.go:1131]: There is an existing test for "parameters in matrix are strings".

**Review by pritidesai — APPROVED** [review]
thank you @EmmaMunley 👍

**Review by lbernick — COMMENTED** [review]
/lgtm

**Comment by jerop** [comment]
/hold 

can we please merge the refactor first? https://github.com/tektoncd/pipeline/pull/6282 🙏🏾 
it'll be harder so sync these changes with the refactor than it would be to update this smaller change

**Comment by jerop** [comment]
/unhold 

https://github.com/tektoncd/pipeline/pull/6282 has been merged

**Comment by pritidesai** [comment]
@jerop it says the changes requested are still pending ⬆️ 

/retest

**Review by jerop — APPROVED** [review]

File: pkg/apis/pipeline/v1/param_types.go
Comment (pritidesai) [inline pkg/apis/pipeline/v1/param_types.go:533]: can we please move this function to `pkg/apis/pipeline/v1/matrix_types.go`?

Comment (pritidesai) [reply pkg/apis/pipeline/v1/param_types.go:533]: along with the other `matrix` members: https://github.com/tektoncd/pipeline/blob/main/pkg/apis/pipeline/v1/matrix_types.go

**Comment by Yongxuanzhang** [comment]
Maybe update the pr description to remove the note if you've finished the work

File: pkg/apis/pipeline/v1/param_types.go
Comment (EmmaMunley) [reply pkg/apis/pipeline/v1/param_types.go]: Yes! Moved both matrix functions

**Comment by EmmaMunley** [comment]
> Maybe update the pr description to remove the note if you've finished the work

@Yongxuanzhang The Matrix Include feature is still in preview mode. This is one of the last 2 PRs for adding validation. :)

File: pkg/apis/pipeline/v1/matrix_types.go
Comment (Yongxuanzhang) [inline pkg/apis/pipeline/v1/matrix_types.go:168]: I'm a little bit confused by this validation, I guess it mainly due to I lack the knowledge of matrix and include. 

Is it possible to use array params in include param and string param for matrix param? Why do we only validate string param for include and array params for matrix?

**Comment by Yongxuanzhang** [comment]
> > Maybe update the pr description to remove the note if you've finished the work
> 
> @Yongxuanzhang The Matrix Include feature is still in preview mode. This is one of the last 2 PRs for adding validation. :)

Oh I thought "This" means this PR, 😂

File: pkg/apis/pipeline/v1/matrix_types.go
Comment (Yongxuanzhang) [reply pkg/apis/pipeline/v1/matrix_types.go:168]: Maybe just add a comment here to explain

Comment (EmmaMunley) [reply pkg/apis/pipeline/v1/matrix_types.go]: Yes, that is true! Sorry for the confusion, added a comment to specify type required.

**Review by QuanZhang-William — COMMENTED** [review]
Thanks @EmmaMunley !

/lgtm

**Review by lbernick — COMMENTED** [review]
/lgtm

**Pull request merged**
